src/Symfony/Component/Security/Http/Authentication/AuthenticatorManager.php

```diff
@@ -0,0 +1,257 @@
+            }
+
+            // authenticate the credentials (e.g. check password)
+            $token = $this->authenticateViaAuthenticator($authenticator, $credentials);
+
+            if (null !== $this->logger) {
+                $this->logger->info('Authenticator successful!', ['token' => $token, 'authenticator' => \get_class($authenticator)]);
+            }
+
+            // success! (sets the token on the token storage, etc)
+            $response = $this->handleAuthenticationSuccess($token, $request, $authenticator);
+            if ($response instanceof Response) {
+                return $response;
+            }
+
+            if (null !== $this->logger) {
+                $this->logger->debug('Authenticator set no success response: request continues.', ['authenticator' => \get_class($authenticator)]);
+            }
+
+            return null;
+        } catch (AuthenticationException $e) {
+            // oh no! Authentication failed!
+            $response = $this->handleAuthenticationFailure($e, $request, $authenticator);
+            if ($response instanceof Response) {
+                return $response;
+            }
+
+            return null;
+        }
+    }
+
+    private function authenticateViaAuthenticator(AuthenticatorInterface $authenticator, $credentials): TokenInterface
+    {
+        // get the user from the Authenticator
+        $user = $authenticator->getUser($credentials);
+        if (null === $user) {
+            throw new UsernameNotFoundException(sprintf('Null returned from "%s::getUser()".', \get_class($authenticator)));
+        }
+
+        $event = new VerifyAuthenticatorCredentialsEvent($authenticator, $credentials, $user);
+        $this->eventDispatcher->dispatch($event);
+        if (true !== $event->areCredentialsValid()) {
+            throw new BadCredentialsException(sprintf('Authentication failed because "%s" did not approve the credentials.', \get_class($authenticator)));
+        }
+
+        // turn the UserInterface into a TokenInterface
+        $authenticatedToken = $authenticator->createAuthenticatedToken($user, $this->providerKey);
+
+        if (true === $this->eraseCredentials) {
+            $authenticatedToken->eraseCredentials();
+        }
+
+        if (null !== $this->eventDispatcher) {
+            $this->eventDispatcher->dispatch(new AuthenticationSuccessEvent($authenticatedToken), AuthenticationEvents::AUTHENTICATION_SUCCESS);
+        }
+
+        return $authenticatedToken;
+    }
+
+    private function handleAuthenticationSuccess(TokenInterface $authenticatedToken, Request $request, AuthenticatorInterface $authenticator): ?Response
+    {
+        $this->tokenStorage->setToken($authenticatedToken);
+
+        $response = null;
+        if ($authenticator instanceof InteractiveAuthenticatorInterface && $authenticator->isInteractive()) {
+            $loginEvent = new InteractiveLoginEvent($request, $authenticatedToken);
```

Comment:
I noticed that, in the new system, this interface is implemented for `form_login`, `json_login`, `http_basic`

But in the OLD system, `http_basic` does NOT dispatch this event, but the following DO: `remote_user`, `x509`, `remember_me`.

This is a mess :). The docs about this event state:

>      * The INTERACTIVE_LOGIN event occurs after a user has actively logged
>     * into your website. It is important to distinguish this action from
>     * non-interactive authentication methods, such as:
>     *   - authentication based on your session.
>     *   - authentication using a HTTP basic or HTTP digest header.

The fact that this event triggered on (for example) `remember_me` and `remote_user` but NOT on `http_basic` seems almost random to me.

Let's fix this `INTERACTIVE_LOGIN` event so that it's triggered from the same auth mechanisms as before. But I would consider this more of a BC layer. The new `LoginSuccessEvent` below is MUCH more powerful than both.

Reply:
Following up on this, I still stand by my statement above. BUT, if someone *did* want to register a listener that was only called during a "real login" (e.g. `form_login`, but not `anonymous` or `remember_me`), how would they do that? The best I can see is that, in their listener, they would:

A) First check to make sure there is a User (to eliminate the anonymous case). 
B) Then check to see if the authenticator is an instanceof `RememberMeAuthenticator` to rule that out

That seems *not* quite ideal... but I'm not sure what to do better. It's almost like this `isInteractive()` WAS a good thing, but it was implemented incorrectly (being added to remember me but not http_basic). Though, `isInteractive()` - even if we tried to implement a new flag "correctly", is still subjective, so maybe there is nothing to do.